substrate/frame/support/procedural/src/construct_runtime_v2/expand/mod.rs

```diff
@@ -0,0 +1,289 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::construct_runtime_v2::{
+	parse::pallets::{
+		AllPalletsDeclaration, ExplicitAllPalletsDeclaration, ImplicitAllPalletsDeclaration,
+	},
+	Def,
+};
+use cfg_expr::Predicate;
+use frame_support_procedural_tools::{
+	generate_crate_access, generate_crate_access_2018, generate_hidden_includes,
+};
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use std::collections::HashSet;
+use syn::{Ident, Result};
+
+use crate::construct_runtime::{check_pallet_number, expand};
+
+use crate::construct_runtime::{
+	decl_all_pallets, decl_integrity_test, decl_pallet_runtime_setup, decl_static_assertions,
+};
+
+use super::parse::runtime_types::RuntimeType;
+
+/// The fixed name of the system pallet.
+const SYSTEM_PALLET_NAME: &str = "System";
+
+pub fn expand(def: Def) -> proc_macro2::TokenStream {
+	let input = def.input;
+
+	let res = match def.pallets {
+		AllPalletsDeclaration::Implicit(ref decl) =>
+			check_pallet_number(input.clone(), decl.pallet_count)
+				.and_then(|_| construct_runtime_implicit_to_explicit(input.into(), decl.clone())),
+		AllPalletsDeclaration::Explicit(ref decl) => check_pallet_number(input, decl.pallets.len())
+			.and_then(|_| {
+				construct_runtime_final_expansion(
+					def.runtime_struct.ident.clone(),
+					decl.clone(),
+					def.runtime_types.clone(),
+				)
+			}),
+	};
+
+	let res = res.unwrap_or_else(|e| e.to_compile_error());
+
+	let res = expander::Expander::new("construct_runtime")
+		.dry(std::env::var("FRAME_EXPAND").is_err())
+		.verbose(true)
+		.write_to_out_dir(res)
+		.expect("Does not fail because of IO in OUT_DIR; qed");
+
+	res.into()
+}
+
+fn construct_runtime_implicit_to_explicit(
+	input: TokenStream2,
+	definition: ImplicitAllPalletsDeclaration,
+) -> Result<TokenStream2> {
+	let frame_support = generate_crate_access_2018("frame-support")?;
+	let mut expansion = quote::quote!(
+		#[frame_support::construct_runtime_v2]
+		#input
+	);
+	for pallet in definition.pallet_decls.iter() {
+		let pallet_path = &pallet.path;
+		let pallet_name = &pallet.name;
+		let pallet_instance = pallet.instance.as_ref().map(|instance| quote::quote!(<#instance>));
+		expansion = quote::quote!(
+			#frame_support::__private::tt_call! {
```

Comment:
curious about whether you contemplated using macro_magic instead of tt_call here and/or whether there were any difficulties preventing that. If it's just because you were mirroring the original setup of `construct_runtime!` that's fine just curious whether there were any barriers I can address on the macro_magic side if applicable.

Reply:
Yes, basically mirroring the previous syntax for expansion in order to minimise the changes. Happy to explore `macro_magic` if you think it comes with some added benefits here.